Refactor kernel installation and add conda recipes

danielfrg opened the pull request:
This PR includes a small change to move the kernel installation part from `setup.py` to another script that has to be called after installation. This is a little annoying but its how most external kernels work. see: [octave kernel](https://github.com/Calysto/octave_kernel/blob/master/octave_kernel/install.py), [scilab_kernel](https://github.com/Calysto/scilab_kernel/blob/master/scilab_kernel/install.py) and [Jupyter extensions](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/install.html#enabling-disabling-extensions).

So to install correctly the SAS kernel now you would need to do:

```
pip install sas_kernel
python -m sas_kernel.install
```

This split also allows `sas_kernel` to be installed correctly in conda environments because the `kernel.json` inside an environment will need the path to that environment python binary.

There is a way to keep the current functionality of installing the kernel with `pip` and also add the fix for conda by refactoring `setup.py` and I can work on that if you guys want but for now I kept it the way most [Jupyter extensions work](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/install.html#enabling-disabling-extensions).

The PR also includes 2 conda recipes for `saspy` and `sas_kernel`. My hope for those recipes is to make them available on [conda-forge](https://github.com/conda-forge) so more people in the community can get them but I wanted to see if someone wants to test them before I move them there.

danielfrg commented on conda.recipe/saspy/meta.yaml:
```diff
@@ -0,0 +1,35 @@
+{%set name = "saspy" %}
```
I committed the recipes for now in case you wanted to test them. After the other changes on this PR are ok I was planning on removing the recipes from the PR and commit them to [conda-forge](https://github.com/conda-forge/) where me and other can maintain them.

jld23 commented:
When I try to run the recipe I get the following error:
What is the correct what to invoke this?
```
$ conda install sas_kernel --use-local
Fetching package metadata ...........

PackageNotFoundError: Packages missing in current channels:
            
  - sas_kernel

We have searched for the packages in the following channels:
            
  - https://conda.anaconda.org/anaconda-fusion/osx-64
  - https://conda.anaconda.org/anaconda-fusion/noarch
  - https://repo.continuum.io/pkgs/free/osx-64
  - https://repo.continuum.io/pkgs/free/noarch
  - https://repo.continuum.io/pkgs/r/osx-64
  - https://repo.continuum.io/pkgs/r/noarch
  - https://repo.continuum.io/pkgs/pro/osx-64
  - https://repo.continuum.io/pkgs/pro/noarch
```

danielfrg commented:
Have you built the packages?

First you need to install `conda-build` go to the `conda.recipe` directory on this PR and build both packages. This should work:

```
conda install conda-build

cd sas_kernel/conda.recipe
conda build saspy sas_kernel

conda install sas_kernel --use-local
```

After that if you conda list you should see `saspy` and `sas_kernel`:

```
$ conda list | grep sas

sas_kernel                2.1.0                    py36_0    local
saspy                     1.2.2                    py36_0    local
```

jld23 commented:
When I tried to build this on my MBP I get the following build error:
conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {'metakernel’}
Not sure how to proceed. There doesn’t appear to be a conda package for metakernel.

Let me know if you need the full traceback.

Jared
> On Aug 17, 2017, at 10:30 AM, Daniel Rodriguez <notifications@github.com> wrote:
> 
> Have you built the packages?
> 
> First you need to install conda-build go to the conda.recipe directory on this PR and build both packages. This should work:
> 
> conda install conda-build
> 
> cd sas_kernel/conda.recipe
> conda build saspy sas_kernel
> 
> conda install sas_kernel --use-local
> After that if you conda list you should see saspy and sas_kernel:
> 
> $ conda list | grep sas
> 
> sas_kernel                2.1.0                    py36_0    local
> saspy                     1.2.2                    py36_0    local
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/sassoftware/sas_kernel/pull/43#issuecomment-323090848>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ADA5yyiqCBF_EPFrC9VKsYEkGaBudKWTks5sZE5_gaJpZM4O4Nxa>.
>

danielfrg commented:
Can you add `conda-forge` as a channel? and then try again.

```
conda config --add channels conda-forge
```

jld23 commented:
@danielfrg after considering the change we definitely want to have a conda package but don't want to give up the single command install. Can you restructure this PR to accomplish that as you indicated in the initial comments.

Thanks!

danielfrg commented:
Definitely.

I added back the kernel installation to the `setup.py` so now it will work as it was working before and install kernel on `pip install`, so single command install. The only difference is that I added a flag for that to be optional so we can disable that when building the conda pkg.

Let me know what you think and if you can test the conda pkgs. If not is ok, I will just go ahead and move the conda recipes out of the PR to the external repo so we can merge this one.

tomweber-sas commented:
Ok, I tried this out too, on a linux VM. I installed anaconda3, installed the conda-build, built sas_kernel/saspy; actually I only said build sas_kernel and it saw the dependancy of saspy and found it local and built it first, then sas_kernel. So that worked too! Cool.
The saspy version it installed was a way out of date version. I see there's a hard coded version in the yaml file.
So, how is this all supposed to work going forward? Where does the code go? Who builds these packages as we push code, increase version numbers and build new pip packages?
pip we do ourselves. Do we build new conda packages at that same time? 

Thanks!
Tom

danielfrg commented:
Yes, the recipe versions are hard coded, there is no way around that since the conda pkg are build for one version.

Going forward I am moving the recipes on this PR to [conda-forge](https://github.com/conda-forge) which is a community maintained repository of conda packages (almost 3000 pkgs right now). Each recipe gets its own git repo on github, called a feedstock, for example this one for [tensorflow](https://github.com/conda-forge/tensorflow-feedstock).

Each repo (feedstock) has different maintainers that update the recipe when new versions are available from upstream then the conda-forge system will automatically build new conda packages for different python versions and OSes (win, macox, linux).

I will do the submission of this recipes for conda-forge and maintain this recipes, I will be more than happy to add people who are interested on maintaining `sas_py` and `sas_kernel`.

danielfrg commented:
I just removed the recipes from the PR, let me know if there is something else I need to do to finish the PR. After that all I would need is a new version of `sas_kernel` on pypi to build the packages on conda-forge.

tomweber-sas commented:
Ok, so, as I understand it, the conda recipes will live in conda forge (whereever, not in saspy or sas_kernel repos). And all we need to do with those is update the versions in the coresponding yaml files when our repo versions change (when we build new pip packages).
Jaren and I would be the only ones doing that. We don't mind if we just need to do a PR from those repos to update the versions and you merge them in, or if you want us to have access the change them. Seem fine for us to just PR a new number.

Then, in sas_kernel, there a new install file and changes to the setup file. Have we verified the changes have no effect on the regular pip install? I just tried doing the conda and it worked but I haven't tried this code using pip (and pypitest to be sure the pip package works too)?
Assuming it's all fine, would there not need to be the same setup/install file changes for saspy too?
saspy is it's own package. saskernel depends on saspy being there, but not the other way around.
So does it need tweaks to be conda installed on it's own when the kernel isn't being installed?

BTW, I'm planning a building a new pip (2.1.7) for saspy some time next week when I finish a couple more issues and add doc for seom new features. We can build a new sas_kernel pip too, after merging this in. I think it's still at 2.1.0, so it'll either be 2.1.1 or we can skip it up to 2.1.7 also, just to have a common starting point for both with the new conda integration.  

Thanks,
Tom

danielfrg commented:
Cool, thats sounds good to me. I will add you guys as maintainers of those when i submit them to conda-forge.

I have tested that the pip installation of this patch and it works as it was working before, feel free to test it.

No, we don't need the changes on `saspy`, the changes here were about the kernel installation and saspy is just a regular python package that doesn't add anything extra.

The timing sounds great.

tomweber-sas commented:
Ok, cool. That sounds good. I'll post back next week when we have new pip's built.

Thanks!
Tom

tomweber-sas commented:
Ok, I've build new pip's for both saspy and sas_kernel. Both are now at version 2.1.7.
Now, you can work your magic and build conda installs from that.

danielfrg commented:
Awesome I will do that!
Thanks so much.

danielfrg commented:
Just a head up conda pkgs are now available:

- https://github.com/conda-forge/sas_kernel-feedstock

tomweber-sas commented:
Awesome! Thanks.

tomweber-sas commented:
@danielfrg I've created a new saspy pypi package: 2.1.8. Can you create the corresponding new conda installer?
sas_kernel hasn't changed, just saspy.
Thanks,
Tom

danielfrg commented:
Yes, I can handle that. Thanks for the heads up.

tomweber-sas commented:
@danielfrg Just created version 2.2.2 of saspy. Want to make the conda version?
Thanks!
Tom

tomweber-sas commented:
scratch that. 2.2.3 in a few minutes. broke something; gotta fixsorry,
tom

danielfrg commented:
No problem, updating the conda pkg to 2.2.3 here: https://github.com/conda-forge/saspy-feedstock/pull/3

If its easier next time you can just create an issue in that repo and ping me there.

tomweber-sas commented:
Thanks! I will do that next time, I mean after this. I built 2.2.3 out of the wrong directory. Bad day. Can you delete that one and do 2.2.4. Man, I'm sorry about this.
Thanks!
Tom

danielfrg commented:
No problem at all, let me know when its ready and I will build 2.2.4

tomweber-sas commented:
Great, thanks! It's out there
Tom

danielfrg commented:
Great, I just updated the conda package to 2.2.4!

tomweber-sas commented:
Awesome, thanks man!
